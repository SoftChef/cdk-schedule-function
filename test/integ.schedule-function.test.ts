import {
  Template,
} from 'aws-cdk-lib/assertions';
import {
  Function,
  InlineCode,
  Runtime,
} from 'aws-cdk-lib/aws-lambda';
import {
  App,
  Duration,
  Stack,
} from 'aws-cdk-lib/core';
import {
  ScheduleFunction,
} from '../src/index';

const fnGetAttArn = (arn: string): { [key: string]: string[] } => {
  return {
    'Fn::GetAtt': [arn, 'Arn'],
  };
};
const fnJoin = (delimiter: string, values: any[]): { [key: string]: any[] } => {
  return {
    'Fn::Join': [delimiter, values],
  };
};
const ref = (name: string): { [key: string]: string } => {
  return {
    Ref: name,
  };
};

const expectedRoles = {
  dispatchTargetFunctionRole: 'ScheduleFunctionDispatchTargetFunctionServiceRole89FC967C',
  createScheduleFunctionRole: 'ScheduleFunctionCreateScheduleFunctionServiceRole5CB97B76',
  updateScheduleFunctionRole: 'ScheduleFunctionUpdateScheduleFunctionServiceRoleCAC6BD24',
  listScheduleFunctionRole: 'ScheduleFunctionListScheduleFunctionServiceRole5EA5341C',
  fetchScheduleFunctionRole: 'ScheduleFunctionFetchScheduleFunctionServiceRole0891609B',
  deleteScheduleFunctionRole: 'ScheduleFunctionDeleteScheduleFunctionServiceRoleE5D76AB9',
};

const expected = {
  dispatchTargetFunctionTimeout: Duration.seconds(30),
  recentMinutes: Duration.minutes(30),
  lambdaFunctionRuntime: 'nodejs14.x',
  scheduleTableName: 'ScheduleFunctionScheduleTable60883DB8',
  scheduleTableQueryByScheduledAtIndexName: 'ScheduleFunctionScheduleTable60883DB8/index/Query-By-ScheduledAt',
  testTargetFunctionName: 'TestTarget4042A7F7',
};

test('minimal usage', () => {
  const app = new App();
  const stack = new Stack(app, 'demo-stack');
  const testTargetFunction = new Function(stack, 'TestTarget', {
    runtime: Runtime.NODEJS_12_X,
    handler: 'index.handler',
    code: new InlineCode(`
      export async function handler() {
        return {
          success: true,
          result: event.context,
        };
      }
    `),
  });
  const scheduleFunction = new ScheduleFunction(stack, 'ScheduleFunction', {
    dispatchTargetFunctionTimeout: expected.dispatchTargetFunctionTimeout,
    recentMinutes: expected.recentMinutes,
  });
  scheduleFunction.addTargetFunction('TestTarget', {
    targetFunction: testTargetFunction,
  });
  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
  template.hasResourceProperties('AWS::Lambda::Function', {
    Runtime: expected.lambdaFunctionRuntime,
    Timeout: expected.dispatchTargetFunctionTimeout.toSeconds(),
  });
  template.hasResourceProperties('AWS::Events::Rule', {
    ScheduleExpression: 'rate(1 minute)',
    State: 'ENABLED',
  });
  template.resourceCountIs('AWS::Lambda::Function', 7);
  template.resourceCountIs('AWS::Lambda::Permission', 1);
  template.resourceCountIs('AWS::IAM::Role', 7);
  template.resourceCountIs('AWS::IAM::Policy', 6); // Test target only has Role, non-policy
  template.resourceCountIs('AWS::DynamoDB::Table', 1);
  // Dispatch target function
  template.hasResourceProperties('AWS::Lambda::Function', {
    Runtime: expected.lambdaFunctionRuntime,
    Role: fnGetAttArn(expectedRoles.dispatchTargetFunctionRole),
    Environment: {
      Variables: {
        SCHEDULE_TABLE_NAME: ref(expected.scheduleTableName),
      },
    },
  });
  template.hasResourceProperties('AWS::IAM::Policy', {
    Roles: [
      ref(expectedRoles.dispatchTargetFunctionRole),
    ],
    PolicyDocument: {
      Statement: [
        {
          Action: [
            'dynamodb:Query',
            'dynamodb:UpdateItem',
          ],
          Effect: 'Allow',
          Resource: [
            fnGetAttArn(expected.scheduleTableName),
            fnJoin('', [
              fnGetAttArn(expected.scheduleTableName),
              '/index/Query-By-ScheduledAt',
            ]),
          ],
        },
        {
          Action: 'lambda:InvokeFunction',
          Effect: 'Allow',
          Resource: [
            fnGetAttArn(expected.testTargetFunctionName),
          ],
        },
      ],
    },
  });
  // Create schedule function
  template.hasResourceProperties('AWS::Lambda::Function', {
    Runtime: expected.lambdaFunctionRuntime,
    Role: fnGetAttArn(expectedRoles.createScheduleFunctionRole),
    Environment: {
      Variables: {
        SCHEDULE_TABLE_NAME: ref(expected.scheduleTableName),
      },
    },
  });
  template.hasResourceProperties('AWS::IAM::Policy', {
    Roles: [
      ref(expectedRoles.createScheduleFunctionRole),
    ],
    PolicyDocument: {
      Statement: [
        {
          Action: 'dynamodb:BatchWriteItem',
          Effect: 'Allow',
          Resource: fnGetAttArn(expected.scheduleTableName),
        },
      ],
    },
  });
  // Update schedule function
  template.hasResourceProperties('AWS::Lambda::Function', {
    Runtime: expected.lambdaFunctionRuntime,
    Role: fnGetAttArn(expectedRoles.updateScheduleFunctionRole),
    Environment: {
      Variables: {
        SCHEDULE_TABLE_NAME: ref(expected.scheduleTableName),
      },
    },
  });
  template.hasResourceProperties('AWS::IAM::Policy', {
    Roles: [
      ref(expectedRoles.updateScheduleFunctionRole),
    ],
    PolicyDocument: {
      Statement: [
        {
          Action: [
            'dynamodb:GetItem',
            'dynamodb:UpdateItem',
          ],
          Effect: 'Allow',
          Resource: fnGetAttArn(expected.scheduleTableName),
        },
      ],
    },
  });
  // List schedules function
  template.hasResourceProperties('AWS::Lambda::Function', {
    Runtime: expected.lambdaFunctionRuntime,
    Role: fnGetAttArn(expectedRoles.listScheduleFunctionRole),
    Environment: {
      Variables: {
        SCHEDULE_TABLE_NAME: ref(expected.scheduleTableName),
      },
    },
  });
  template.hasResourceProperties('AWS::IAM::Policy', {
    Roles: [
      ref(expectedRoles.listScheduleFunctionRole),
    ],
    PolicyDocument: {
      Statement: [
        {
          Action: [
            'dynamodb:Query',
            'dynamodb:Scan',
          ],
          Effect: 'Allow',
          Resource: fnGetAttArn(expected.scheduleTableName),
        },
      ],
    },
  });
  // Fetch schedule function
  template.hasResourceProperties('AWS::Lambda::Function', {
    Runtime: expected.lambdaFunctionRuntime,
    Role: fnGetAttArn(expectedRoles.fetchScheduleFunctionRole),
    Environment: {
      Variables: {
        SCHEDULE_TABLE_NAME: ref(expected.scheduleTableName),
      },
    },
  });
  template.hasResourceProperties('AWS::IAM::Policy', {
    Roles: [
      ref(expectedRoles.fetchScheduleFunctionRole),
    ],
    PolicyDocument: {
      Statement: [
        {
          Action: 'dynamodb:GetItem',
          Effect: 'Allow',
          Resource: fnGetAttArn(expected.scheduleTableName),
        },
      ],
    },
  });
  // Delete schedule function
  template.hasResourceProperties('AWS::Lambda::Function', {
    Runtime: expected.lambdaFunctionRuntime,
    Role: fnGetAttArn(expectedRoles.deleteScheduleFunctionRole),
    Environment: {
      Variables: {
        SCHEDULE_TABLE_NAME: ref(expected.scheduleTableName),
      },
    },
  });
  template.hasResourceProperties('AWS::IAM::Policy', {
    Roles: [
      ref(expectedRoles.deleteScheduleFunctionRole),
    ],
    PolicyDocument: {
      Statement: [
        {
          Action: [
            'dynamodb:GetItem',
            'dynamodb:DeleteItem',
          ],
          Effect: 'Allow',
          Resource: fnGetAttArn(expected.scheduleTableName),
        },
      ],
    },
  });
  // Schedule Table
  template.hasResourceProperties('AWS::DynamoDB::Table', {
    AttributeDefinitions: [
      {
        AttributeName: 'scheduleId',
        AttributeType: 'S',
      },
      {
        AttributeName: 'scheduledAt',
        AttributeType: 'S',
      },
      {
        AttributeName: 'targetType',
        AttributeType: 'S',
      },
      {
        AttributeName: 'targetId',
        AttributeType: 'S',
      },
    ],
    KeySchema: [
      {
        AttributeName: 'scheduleId',
        KeyType: 'HASH',
      },
    ],
    ProvisionedThroughput: {
      ReadCapacityUnits: 1,
      WriteCapacityUnits: 1,
    },
    GlobalSecondaryIndexes: [
      {
        IndexName: 'Query-By-ScheduledAt',
        KeySchema: [
          {
            AttributeName: 'scheduledAt',
            KeyType: 'HASH',
          },
        ],
        Projection: {
          ProjectionType: 'ALL',
        },
        ProvisionedThroughput: {
          ReadCapacityUnits: 1,
          WriteCapacityUnits: 1,
        },
      },
      {
        IndexName: 'Query-By-TargetType',
        KeySchema: [
          {
            AttributeName: 'targetType',
            KeyType: 'HASH',
          },
        ],
        Projection: {
          ProjectionType: 'ALL',
        },
        ProvisionedThroughput: {
          ReadCapacityUnits: 1,
          WriteCapacityUnits: 1,
        },
      },
      {
        IndexName: 'Query-By-TargetId',
        KeySchema: [
          {
            AttributeName: 'targetId',
            KeyType: 'HASH',
          },
        ],
        Projection: {
          ProjectionType: 'ALL',
        },
        ProvisionedThroughput: {
          ReadCapacityUnits: 1,
          WriteCapacityUnits: 1,
        },
      },
    ],
  });
});