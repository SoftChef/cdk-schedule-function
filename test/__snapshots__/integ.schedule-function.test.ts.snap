// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minimal usage 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "ScheduleFunctionCreateScheduleFunction839A34E1": Object {
      "DependsOn": Array [
        "ScheduleFunctionCreateScheduleFunctionServiceRole5CB97B76",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "f9a636f2c7218f5d870b47cc0e27fcaa7c58bd800d75cb19df2fbe9381e65094.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "RECENT_MINUTES": "30",
            "SCHEDULE_TABLE_NAME": Object {
              "Ref": "ScheduleFunctionScheduleTable60883DB8",
            },
            "TARGET_FUNCTIONS_NAME": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "{\\"TestTarget\\":\\"",
                  Object {
                    "Ref": "TestTarget4042A7F7",
                  },
                  "\\"}",
                ],
              ],
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ScheduleFunctionCreateScheduleFunctionServiceRole5CB97B76",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ScheduleFunctionCreateScheduleFunctionServiceRole5CB97B76": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ScheduleFunctionDeleteScheduleFunctionC18DF41E": Object {
      "DependsOn": Array [
        "ScheduleFunctionDeleteScheduleFunctionServiceRoleE5D76AB9",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "d8ad11930a5726df5f40b07d4cebc9a2afb3e303978c92b234d45df3fcc6d3dc.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SCHEDULE_TABLE_NAME": Object {
              "Ref": "ScheduleFunctionScheduleTable60883DB8",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ScheduleFunctionDeleteScheduleFunctionServiceRoleE5D76AB9",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ScheduleFunctionDeleteScheduleFunctionServiceRoleE5D76AB9": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ScheduleFunctionDispatchTargetFunction6A796318": Object {
      "DependsOn": Array [
        "ScheduleFunctionDispatchTargetFunctionServiceRole89FC967C",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "3ccfc3b3f62259322ae6d9d65849ba60e580c38d33522259d522ad4dc94ba83e.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SCHEDULE_TABLE_NAME": Object {
              "Ref": "ScheduleFunctionScheduleTable60883DB8",
            },
            "TARGET_FUNCTIONS_NAME": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "{\\"TestTarget\\":\\"",
                  Object {
                    "Ref": "TestTarget4042A7F7",
                  },
                  "\\"}",
                ],
              ],
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ScheduleFunctionDispatchTargetFunctionServiceRole89FC967C",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ScheduleFunctionDispatchTargetFunctionServiceRole89FC967C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ScheduleFunctionDispatchTargetRuleAllowEventRuledemostackScheduleFunctionDispatchTargetFunction046F149B139F2B36": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "ScheduleFunctionDispatchTargetFunction6A796318",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "ScheduleFunctionDispatchTargetRuleE8DBE1C8",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ScheduleFunctionDispatchTargetRuleE8DBE1C8": Object {
      "Properties": Object {
        "ScheduleExpression": "rate(1 minute)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "ScheduleFunctionDispatchTargetFunction6A796318",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "ScheduleFunctionFetchScheduleFunctionD70172FC": Object {
      "DependsOn": Array [
        "ScheduleFunctionFetchScheduleFunctionServiceRole0891609B",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "7d271538c6a81576bf2b5e4ee13f7d65d328f3fee09bc673028d1e39d118828d.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SCHEDULE_TABLE_NAME": Object {
              "Ref": "ScheduleFunctionScheduleTable60883DB8",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ScheduleFunctionFetchScheduleFunctionServiceRole0891609B",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ScheduleFunctionFetchScheduleFunctionServiceRole0891609B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ScheduleFunctionListScheduleFunction45B7EF79": Object {
      "DependsOn": Array [
        "ScheduleFunctionListScheduleFunctionServiceRole5EA5341C",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "ddce7a914ca964e6ac235eae162bc045d43963ac4ae4f3febf3c80aacb721536.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SCHEDULE_TABLE_NAME": Object {
              "Ref": "ScheduleFunctionScheduleTable60883DB8",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ScheduleFunctionListScheduleFunctionServiceRole5EA5341C",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ScheduleFunctionListScheduleFunctionServiceRole5EA5341C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ScheduleFunctionScheduleTable60883DB8": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "scheduleId",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "scheduledAt",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "targetType",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "targetId",
            "AttributeType": "S",
          },
        ],
        "GlobalSecondaryIndexes": Array [
          Object {
            "IndexName": "Query-By-ScheduledAt",
            "KeySchema": Array [
              Object {
                "AttributeName": "scheduledAt",
                "KeyType": "HASH",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
            "ProvisionedThroughput": Object {
              "ReadCapacityUnits": 1,
              "WriteCapacityUnits": 1,
            },
          },
          Object {
            "IndexName": "Query-By-TargetType",
            "KeySchema": Array [
              Object {
                "AttributeName": "targetType",
                "KeyType": "HASH",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
            "ProvisionedThroughput": Object {
              "ReadCapacityUnits": 1,
              "WriteCapacityUnits": 1,
            },
          },
          Object {
            "IndexName": "Query-By-TargetId",
            "KeySchema": Array [
              Object {
                "AttributeName": "targetId",
                "KeyType": "HASH",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
            "ProvisionedThroughput": Object {
              "ReadCapacityUnits": 1,
              "WriteCapacityUnits": 1,
            },
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "scheduleId",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 1,
          "WriteCapacityUnits": 1,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "ScheduleFunctionUpdateScheduleFunction2DA1CD25": Object {
      "DependsOn": Array [
        "ScheduleFunctionUpdateScheduleFunctionServiceRoleCAC6BD24",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "4ae5d4fc93c8a9e6d9e12195f6390b18f40961ba753e045a9e3f1e161816a7ea.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SCHEDULE_TABLE_NAME": Object {
              "Ref": "ScheduleFunctionScheduleTable60883DB8",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ScheduleFunctionUpdateScheduleFunctionServiceRoleCAC6BD24",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ScheduleFunctionUpdateScheduleFunctionServiceRoleCAC6BD24": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ScheduleFunctiondeleteschedulepolicy0E23AE84": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:GetItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "ScheduleFunctionScheduleTable60883DB8",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ScheduleFunctiondeleteschedulepolicy0E23AE84",
        "Roles": Array [
          Object {
            "Ref": "ScheduleFunctionDeleteScheduleFunctionServiceRoleE5D76AB9",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ScheduleFunctiondispatchschedulepolicy7AE2621D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:Query",
                "dynamodb:UpdateItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "ScheduleFunctionScheduleTable60883DB8",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "ScheduleFunctionScheduleTable60883DB8",
                          "Arn",
                        ],
                      },
                      "/index/Query-By-ScheduledAt",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TestTarget4042A7F7",
                    "Arn",
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ScheduleFunctiondispatchschedulepolicy7AE2621D",
        "Roles": Array [
          Object {
            "Ref": "ScheduleFunctionDispatchTargetFunctionServiceRole89FC967C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ScheduleFunctionfetchschedulepolicy9B976A4F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "dynamodb:GetItem",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "ScheduleFunctionScheduleTable60883DB8",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ScheduleFunctionfetchschedulepolicy9B976A4F",
        "Roles": Array [
          Object {
            "Ref": "ScheduleFunctionFetchScheduleFunctionServiceRole0891609B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ScheduleFunctionlistschedulespolicy36680AA3": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:Query",
                "dynamodb:Scan",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "ScheduleFunctionScheduleTable60883DB8",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ScheduleFunctionlistschedulespolicy36680AA3",
        "Roles": Array [
          Object {
            "Ref": "ScheduleFunctionListScheduleFunctionServiceRole5EA5341C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ScheduleFunctionputschedulepolicy9BCF4420": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "dynamodb:BatchWriteItem",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "ScheduleFunctionScheduleTable60883DB8",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ScheduleFunctionputschedulepolicy9BCF4420",
        "Roles": Array [
          Object {
            "Ref": "ScheduleFunctionCreateScheduleFunctionServiceRole5CB97B76",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ScheduleFunctionupdateschedulepolicy9AC28D81": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:GetItem",
                "dynamodb:UpdateItem",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "ScheduleFunctionScheduleTable60883DB8",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ScheduleFunctionupdateschedulepolicy9AC28D81",
        "Roles": Array [
          Object {
            "Ref": "ScheduleFunctionUpdateScheduleFunctionServiceRoleCAC6BD24",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestTarget4042A7F7": Object {
      "DependsOn": Array [
        "TestTargetServiceRoleED06BFE8",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "
      export async function handler() {
        return {
          success: true,
          result: event.context,
        };
      }
    ",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestTargetServiceRoleED06BFE8",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestTargetServiceRoleED06BFE8": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
